### PR TITLE
[WIP] POC/tr 4047

### PR DESCRIPTION
**Pull request by bziondik**
Related to: https://oat-sa.atlassian.net/browse/TR-4047

## Acceptance criteria:

- AC1: define the pattern a PCI must present to activate the navigation lock feature
- AC2: create a prototype of PCI to showcase the feature (can be a clone of an existing one, ex: likert)
   - AC2a: from the item editor, an option allows to activate/deactivate the pattern
   - AC2b: WHEN the option is activated, THEN the PCI MUST present the pattern
   - AC2c: WHEN the option is deactivated, THEN the PCI MUST NOT present the pattern

# Solution

- add property `navigationLock` to creator widget of PCI check example here https://github.com/oat-sa/extension-tao-itemqti-pci/pull/310/commits/e939ab892eb761668501203d4be0f6d2c8e8fb4d
- trigger `responseChange` event when PCI is answered https://github.com/oat-sa/extension-tao-itemqti-pci/pull/310/commits/2c3a53d75548bd657f39a7790c6507bbda800aee
- check that response has value and default response is null  https://github.com/oat-sa/extension-tao-itemqti-pci/pull/310/commits/968f35770b415a25a13e3b44459e0e0603150015
- POC of navigationLock plugin that define PCI by `pci.Element.properties.navigationLock` https://github.com/oat-sa/extension-tao-itemqti-pci/pull/310/commits/b34cd710e09fb0c774daeca8f064c04ce156e2c7

### CODE REVIEW

File: views/js/pciCreator/dev/likertMy/creator/tpl/propertiesForm.tpl [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+</div>
+<div class="panel">
+    <label>
+        <input name="navigationLock" type="checkbox" {{#if navigationLock}}checked="checked"{{/if}}/>
+        <span class="icon-checkbox"></span>
+        {{__ "Navigation lock"}}
+    </label>
+    <span class="icon-help tooltipstered" data-tooltip="~ .tooltip-content:first" data-tooltip-theme="info"></span>
+    <span class="tooltip-content">
+        {{__ 'If this box is checked the test runner navigation to be disabled until the test-taker gave an answer to it.'}}
```
Comment (jsconan) [inline views/js/pciCreator/dev/likertMy/creator/tpl/propertiesForm.tpl:19]: **quibble:** Not exactly. The RFE-223 proposed several possibilities, and the one that was chosen is not disabling the navigation. Instead, it blocks it, the user can still click the navigation button, but when doing so, a message is displayed and the navigation is canceled.

File: migrations/Version202205130814351465_qtiItemPci.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                'id' => 'navigationlock',
+                'name' => 'Navigation lock',
+                'module' => 'qtiItemPci/plugins/controls/navigation/navigationlock',
+                'bundle' => 'qtiItemPci/loader/testPlugins.min',
+                'description' => 'Disable navigation based on extra requirements',
+                'category' => 'controls',
+                'active' => true,
+                'tags' => [ 'core' ]
+            ]));
+        }
```
Comment (jsconan) [inline migrations/Version202205130814351465_qtiItemPci.php:37]: **quibble:** the plugin was out of scope, but ok, why not

File: views/js/pciCreator/dev/likertMy/runtime/js/likertMy.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            qtiCustomInteractionContext.notifyReady(this);
+
+            //listening to dynamic configuration change
+            this.on('levelchange', function(level){
+                self.config.level = level;
+                renderer.renderChoices(self.id, self.dom, self.config);
+            });
+            $(this.dom).find('input').on('change', function() {
+                // Notify the response has changed
+                self.trigger('responseChange', [self.getResponse()]);
```
Comment (jsconan) [inline views/js/pciCreator/dev/likertMy/runtime/js/likertMy.js:60]: **quibble:** not compliant with the IMS PCI standard, but this is acceptable with respect to the TAO flavor.

File: views/js/plugins/controls/navigation/navigationlock.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+                const itemRunner = testRunner.itemRunner;
+                const elementsOnPage = itemRunner.getData().body.elements;
+
+                const PCINavigationLockExists = {};
+                Object.values(elementsOnPage).forEach(elem => {
+                    if (elem.qtiClass === 'customInteraction' && elem.properties && elem.properties.navigationLock && elem.attributes.responseIdentifier) {
+                        PCINavigationLockExists[elem.attributes.responseIdentifier] = false;
+                    }
+                });
```
Comment (jsconan) [inline views/js/plugins/controls/navigation/navigationlock.js:49]: **thought:** Ok, this shows how to detect the PCI requiring the navigationLock feature. In the end, this must go in a helper as per [TR-4051](https://oat-sa.atlassian.net/browse/TR-4051).

File: views/js/plugins/controls/navigation/navigationlock.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+                if (!Object.keys(PCINavigationLockExists).length) {
+                    return;
+                }
+                testRunner.trigger('disablenav');
```
Comment (jsconan) [inline views/js/plugins/controls/navigation/navigationlock.js:54]: **issue:** this contradicts the solution described in the RFE. It must not disable the navigation. Instead, it must hook it and display the messages defined by [TR-4052](https://oat-sa.atlassian.net/browse/TR-4052), then cancel the action.
This must be done from the ticket [TR-4054](https://oat-sa.atlassian.net/browse/TR-4054), so not a big deal, but be careful to not give a wrong assumption already.

File: views/js/plugins/controls/navigation/navigationlock.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+                itemRunner.on('responsechange', (responses) => {
```
Comment (jsconan) [inline views/js/plugins/controls/navigation/navigationlock.js:56]: **issue:** this also contradicts the solution described in the RFE. We only need to check the response before allowing it to navigate.

**Review by jsconan — COMMENTED** [review]
The draft looks ok with respect to the AC: the pattern to identify the PCI is there, and there is an example of how to build it. It works pretty well. Good job on this part!

However, there is way more than what was expected. Especially the TR plugin, which was out of scope, and that is not applying the solution described in the RFE. Anyway, this can be corrected in the next tickets.

This being said, could you highlight in the code the portions that relate to the pattern and which need to be inserted in other PCIs so that it would ease the documentation writing ([TR-4050](https://oat-sa.atlassian.net/browse/TR-4050))?

And also, this is customer-specific, I don't think it needs to be added to the product. So better to move it to the customer extension ;)

**Comment by bziondik** [comment]
closed in favor https://github.com/oat-sa/extension-tao-depp/pull/333

**Pull request closed without merging**